Stdlib priority queues

backtracking opened the pull request:
Hi OCaml developers,

Now that we have dynamic arrays in the standard library, it makes sense to derive an implementation of mutable priority queues from that and I think it would make a nice addition to the standard library. This is what this PR is about.

I'm aware that the documentation of `Dynarray` already provides a minimal implementation of priority queues, but I'd like to advocate that
- it is a bit of a pity to have to copy-paste such a standard data structure from a documentation;
- some functions are obviously missing, such as peeking, iteration, linear-time building from a set of values.

Beside, we can find around several OCaml software that have their own local implementations of priority queues and that would have obviously benefited from priority queues being readily available from the OCaml standard library.

Notes/points to be discussed:
- The code could be slightly more efficient if `Dynarray.unsafe_get/set` were exported.
- The exception [Empty] is local to the functor output signature. I think this is cleaner than a global exception declared at toplevel, but opinions are welcome.
- `fold` is a `fold_left`-like, to be consistent with module `Queue`.
- There are a few tests in `testsuite/tests/lib-pqueue/test.ml` already, but I can add more.

Comments are welcome.
Please be indulgent, as it is my very first PR to the OCaml code base :wink:

backtracking committed:
stdlib: new module Pqueue

backtracking committed:
a more versatile API

functor Make provides a simple API (with priorities apart
from values) and functor MakePoly provides a generic API
where the type of elements is user-provided

gasche commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,208 @@
+
+{[
+  module IntPQ = Pqueue.MakePoly(struct
+    type _ t = string
```
Why `string` instead of `int`? (The intention of `IntPQ` was to suggest "priority queue of integers".)

backtracking replied:
Oups, a last-minute change typo. I thought that `string` was more interesting than `int`, since `MinQueue` already provides queues of integers. I'll fix that.

gasche reviewed (commented):
The problem with `Dynarray.unsafe_{get,set}` is that they break memory safety in presence of concurrent usage of the data structure. If someone else calls `clear` on your priority queue from another domain, you may end up segfaulting, and we have decided that OCaml code must not do that -- even if its documentation says "please don't perform racy accesses to my structure".

Skimming over your implementation, I see few cases where we could actually have static guarantees that would let us save many bound checks using `Dynarray.unsafe_{get,set}`:

1. The only case I can easily see are the `to_heap` function that call `heapify` on a heap that we know is local (we have not shared it with anyone yet) and thus cannot receive concurrent updates.
2. I suppose that more functions could be safely made to use unsafe accessors, but it would require a different Dynarray API where you can borrow a "fixed-size view" of the array, for a portion of the code that you know will not change the number of elements. (At the low-level this fixed-size view would be the backing array directly, with a check when returning the borrow that no other domain changed the backing array to a new version in the meantime.)

c-cube commented:
> you may end up segfaulting, and we have decided that OCaml code must not do that

I hadn't realized this was valid for `unsafe_…` operations as well, actually. I'm a bit surprised. With something like `Array.unsafe_get` a user can already happily segfault even in the absence of data races; what makes races so special that the user needs to be protected even when trying to use the `unsafe_…` functions?

gasche commented:
@c-cube if the user calls `Dynarray.unsafe_get` directly, they are on their own. But if they call `PQueue.pop_min`, they do not expect a segfault, and this severely restricts the ability to use an `unsafe_` function inside `pop_min` (for example). Many unsafe-introducing optimizations that were valid for OCaml 4 are not valid for OCaml 5.

redianthus commented:
This needs a change entry in the `Standard library` section of the `Changes` file.

c-cube commented:
@gasche thank you, that makes a lot of sense!

backtracking committed:
update Changes

redianthus commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,208 @@
+    in many algorithms, (e.g. when peeking at several priority queues
+    and then selecting one to remove from).
+
+    @since 5.2
```
I believe it's too late for 5.2 (see d46f79c898bf52e3bf76027d4570754916c617e5), it should be 5.3 instead. I'll let @octachron confirm.

Octachron replied:
Indeed, the change would be in 5.3 not 5.2 .

backtracking replied:
changed to 5.3

gasche commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,223 @@
+    val to_seq: 'a t -> (prio * 'a) Seq.t
+  end
+
+module Make(Prio: OrderedType) =
```
I would add a signature constraint here `: S with prio = Prio.t` to make type-checking fail earlier (than when comparing the .mli and .ml together) if S and the functor get out of sync.

backtracking replied:
Good point. Done.

gasche commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,208 @@
+          zero if the priorities [p1] and [p2] are equal,
+          [f p1 p2] is strictly negative if [p1] has higher priority than
+          [p2], and [f p1 p2] is strictly positive if [p1]
+          has lower priority than [p2].
```
I found this paragraph a bit confusing for non-experts. "higher priority" would be understood in informal language as "the thing that must be done first", and you mean exactly the opposite here. Is there a way to reformulate this to avoid this confusion? Or maybe we could point it out explicitly somewhere in the documentation?

**Edit**: no, in fact, "higher priority" here means "selected first". I had read it correctly initially then somehow confused myself. In any case, I think that the fact that "higher" means "smaller" deserves to be formulated carefully.

backtracking replied:
Hi @gasche,
I tried to address these concerns in 3c53d82c, with two different signatures `Min` and `Max` and two different functors `MakeMin` and `MakeMax`. Comments are welcome.

gasche reviewed (commented):
The fact that `MaxQueue` has a function called `pop_min` that pops the... maximum element may be found confusing by non-experts. I wonder if this is unavoidable, and something we could consider pointing out explicitly, or if there are alternative naming conventions (`pop_best`?) that avoid the issue and would be palatable to experts.

Relatedly: I think that there are people who want priority queues but are not sure what "Min-priority" and "Max-priority" mean, would it make sense to give a couple sentence of explanation for non-experts or give a short example to clear up any misunderstanding?

backtracking committed:
give module Make a signature constraint

backtracking committed:
too late for 5.2

backtracking committed:
add Pqueue to stdlib documentation

redianthus commented on testsuite/tests/lib-pqueue/test.ml:
```diff
@@ -0,0 +1,119 @@
+                          | Some x -> assert (x = a.(i))
+  done;
+  assert (is_empty q)
+  )
```
You can replace `PQ.( ... )` by `let open PQ in ...`, it's more readable IMO. (Same on line 90)

yallop replied:
```ocaml
let open Pqueue.MakePoly(E) in
   ...
```

backtracking replied:
Thanks for the suggestion. Done.

redianthus commented on testsuite/tests/lib-pqueue/test.ml:
```diff
@@ -0,0 +1,119 @@
+  for _ = 1 to 4 do
+    let x = Q.min_elt q in assert (x = Q.pop_min q)
+  done;
+  ()
```
You can replace `done; ()` by `done`.

backtracking replied:
Sure. But I find it easier to insert more tests this way.

redianthus commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,223 @@
+      Dynarray.length
+
+    let is_empty h =
+      length h = 0
```
```ocaml
let is_empty =
  Dynarray.is_empty
```

backtracking replied:
Done.

redianthus commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,223 @@
+      Dynarray.get h 0
+
+    let min_elt_opt h =
+      if Dynarray.length h = 0 then None else Some (Dynarray.get h 0)
```
`Dynarray.is_empty` (same above for `min_elt`)

backtracking replied:
Done.

redianthus commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,223 @@
+        let r = Dynarray.get h 0 in
+        sift_down h ~len:(n - 1) 0 x;
+        Some r
+      )
```
I am not sure if this is worth duplicating the code instead of using `try Some (pop_min h) with Empty -> None`. Both are used in the stdlib.

gasche replied:
(Surely you mean `match pop_min h with x -> Some x | exception Empty -> None`.)

jberdine replied:
In general it is a nice property of code that does not raise exceptions as part of its interface to also not clobber the existing exception backtrace. The code as written has this property, but since `pop_min` does not, and should not, use `raise_notrace`, then defining the option variant from the exception variant would sacrifice this property.

redianthus commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,223 @@
+        Some r
+      )
+
+    let remove_min h =
```
Maybe it could be named `drop_min` ? So that we would have `Pqueue.{pop_min,drop_min}` like in `Stack.{pop,drop}`. `Queue` only has `pop`, I opened #12884 to add `drop`.

backtracking committed:
use Dynarray.is_empty

backtracking commented:
> The fact that `MaxQueue` has a function called `pop_min` that pops the... maximum element may be found confusing 

Indeed. One solution would be to have two signatures `MinPQ` and `MaxPQ` (instead of `S`), the former with `remove_min` and such, and the latter with `remove_max` and such. This is for instance what is done in Sedgewick&Wayne's Algorithms (4th edition, page 309) with both [MinPQ](https://algs4.cs.princeton.edu/code/edu/princeton/cs/algs4/MinPQ.java.html) and [MaxPQ](https://algs4.cs.princeton.edu/code/edu/princeton/cs/algs4/MaxPQ.java.html). Note: Their code is duplicated but of course `MaxPQ` implementation can be straightforwardly derived from `MinPQ`'s.

backtracking commented:
Regarding the use of `Dynarray` unsafe operations, I now understand that my request was only meaningful for sequential code. I guess we can forget it.

backtracking committed:
stdlib: better interface for priority queues

introducing two different signatures for min-priority queues and
max-priority queues (and accordingly two functors to build them)
is much clearer, at the cost of a slightly bigger interface for
Pqueue

backtracking committed:
make alldepend

gasche commented:
I would like to come back to this PR as I think it would be a shame to let this wait forever. In principle I think that there is a agreement to have priority queues in the stdlib, but the API design questions are proving tricky.

Re-reading the API after a long pause:
- I like the separation between the `Min` and `Max` interfaces, I think they really add clarity / user-friendliness (at a cost of duplicating some boilerplate internally which I think is acceptable)
- On the other hand, I am not convinced by the `Make / MakePoly` duology anymore. I think that we could use just `MakePoly` (renamed into `Make`), and document the fact that if people cannot easily compute the priority from the item, they should store (priority, item) pairs in the queue. Note that for most functions the API we get this way is the same as the one we currently offer, so there is very little loss of convenience.

Also: today people that could use `MakePoly` might make the mistake of using `Make` instead (by, say, calling `Queue.add event.deadline event queue`) just because they find it first and it looks simpler, getting a slightly more cumbersome interface and a less efficient represetation.

One minor downside of `MakePoly` is that the interface is not the same as the standard `OrderedType` from other modules, so say one cannot apply `MakePoly(Int)` and be done with it. But (1) redefining the correct module from `Int` is trivial, and (2) in fact applying `Make(Int)` does not give the desired interface either -- if you want queues of integers.

dbuenzli commented:
I didn't have the time to look at this more closely but in case that may help inform the design, as I already mentioned in the dynarray PR, I have at least six implementations of priority queues in my public code (e.g. in [`gg`](https://github.com/dbuenzli/gg/blob/d4333c8b2abdfc18c01f9c0d15521ce099bbb374/src_kit/gg__pgon2.ml#L29-L66), [`affect`](https://github.com/dbuenzli/affect/blob/5ccf1e2fea197f6257ccebffaff8ca0970a0d86c/src/funix.ml#L53-L96), [`rtime`](https://github.com/dbuenzli/rtime/blob/3a4a7ec3284cf7d2a8d06e0e8d89a663aa2e72c8/src/rtime.ml#L23-L89), [`fut`](https://github.com/dbuenzli/fut/blob/907de63df12815f2df2cb796baa7fa37ac2758d4/src-select/fut_backend.ml#L27-L69), [`useri`](https://github.com/dbuenzli/useri/blob/4646a7b8f3c0d6ce34b1c3f789eebe56cb75b9d2/src-tsdl/useri_backend_tsdl.ml#L56-L109), [`react` (on weak arrays)](https://github.com/dbuenzli/react/blob/aebf35179964e73a7ef557c6a56175c7abdb9947/src/react.ml#L234-L280)). It would be interesting to see how easily they can be switched over to use the proposal here.

gasche commented:
Through the lens of the "Make vs. MakePoly" question (do we expect users to provide priorities separately from values (Make), or values with a comparison function (MakePoly)), it looks like `gg`, `affect`, `rtime` and `react` are clearly MakePoly users, while `fut` and `useri` offer a more Make-like interface where the `add` functions expects a priority/time/deadline first and an element second. This coincides with whether the "pair of a priority with other stuff" is exposed as an abstraction of its own, or merely an implementation detail. For example `rtime` and `fut` use very similar `deadline` types, but in `rtime` this is visible to the user as a first-class notion and in particular a deadline can be explicitly cancelled, while in `fut` this is an implementation detail and the rest of the code focuses on "actions" (the values that are paired with a time/deadline in the queue).

This validates that MakePoly is important to have. I don't know if I would conclude anything regarding Make. (My guess is that you would have been fine writing the code slightly differently if MakePoly was the only option, but who knows.)

dbuenzli commented:
I had a quick look at the proposal. Here are a comments:

1. Regarding separate values and priorities, that's what @pqwy's [psq](https://pqwy.github.io/psq/doc/psq/Psq/index.html) does using a two argument `Make` where both are ordered because efficient access to prioritized elements is also provided. It seems containers [`CCHeap`](https://ocaml.org/p/containers/latest/doc/CCHeap/index.html) uses a single argument. I couldn't quickly find a priority queue in either batteries or base.

2. On the fact that `Int` is not readily applicable to `MakePoly`, it's easy to provide an "adapter" functor (see below).

3. I would spell out `prio` as `priority`.
4. I would leave out the two `Int` instances.

Personally I think that except if the API wants to go in the direction of @pqwy's psq, the separate priority and values is not fundamental and can be recovered from `MakePoly` as shown in this very PR. As such I think I would rather have the following signature for `pqueue.mli`

```ocaml
module type OrderedType = sig type t val compare : t -> t -> int end
module type OrderedPolytype = sig type 'a t val compare : 'a t -> 'a t -> int end
module type Min = sig … end
module type Max = sig … end

module MakeMin (Elt : OrderedPolytype) : Min
module MakeMax (Elt : OrderedPolytype) : Max
module MakeOrderedPolytypeOfOrderedType (E : OrderedType) : OrderedPolytype
```

(and alternatively we could forget about the polytype and simply go with `OrderedType` like other functors of the stdlib)

dbuenzli commented:
> (and alternatively we could forget about the polytype and simply go with `OrderedType` like other functors of the stdlib)

Thinking more about it, somehow `OrderedPolytype` just saves you if your type has a single parameter. The problem remains if you have more than one. I think we should rather go as is usual in the stdlib (if evidence serves, none of my use cases needed a polymorphic type).

gasche commented:
I disagree with your comments on the "poly" interface, but this is a bit tricky to explain clearly.

There are two interfaces one could consider for queues:
- separated priorities: the priorities are provided separately from the elements
- internal priorities: the priorities are stored in the elements

```ocaml
module type OrderedType = Set.OrderedType

module type SeparatedPriorityQueue = functor (Prio : OrderedType) -> sig
  type 'a t
  val add : Prio.t * 'a -> 'a t -> unit
end

module type InternalPriorityQueue = functor (Elem : OrderedType) -> sig
  type t
  val add : Elem.t -> t -> unit
end
```

For internal priority queues, the natural way to define them is to use the elements themselves as priorities -- as ordered types. But notice that you then have a *monomorphic* queue as a result, that can only store elements of one type. This is inconvenient in practice, for the same reason that people prefer to work with `'a list` than with `Set(Something).t`.

If you want *polymorphic* internal priority queues, you need a functor over `OrderedPolyType` instead.

```ocaml
module type InternalPolyPriorityQueue = functor (Elem : OrderedPolyType) -> sig
  type 'a t
  val add : 'a Elem.t -> t -> unit
end
```

This is strictly more expressive, and in particular you cannot define `InternalPolyPriorityQueue` on top of `SeparatedPriorityQueue`, but the converse is true, `SeparatedPriorityQueue(Prio)` is just `InternalPolyPriorityQueue(struct type 'a t = Prio.t * 'a end)`. (You can also define a `SeparatedPolyPriorityQueue` where priorities themselves have a parameter.)

gasche commented:
In other words: we want to cover both "separated priority" and "internal priority" API expectations, *and* to have the functor a type of queue that is polymorphic over the type of elements stored in the queue. The only way to satisfy both needs is to use an OrderedPolyType.

You point out that your own use-cases do not use queues polymorphic over the type of elements. So you would be content with the first, monomorphic `InternalPriorityQueue` signature I sketched above.
Personally I agree that exposing this form of polymorphism is useful, even if it makes for a slightly more complex API.

dbuenzli commented:
What I'm saying is that internal priorities with monomorphic types will likely cover 95% of your cases and is good enough for the `Stdlib`. You can always add more complexity in an API to try to cover all of the cases but it's not necessarily desirable/and or natural[^1].

[^1]: E.g. my first reaction to `OrderedPolytype` was: oh but you are forced to use polymorphic equality to use it, you don't have a comparison function to compare the alphas, this is not good. Of course that's not the case but I don't find this interface to be natural.

dbuenzli commented:
> In other words: we want to cover both "separated priority" and "internal priority" API expectations,

My gut feeling is that it's a bit of featuritis. But if you really want this I would rather suggest to have separate interfaces rather than the confusing `OrderedPolyType` and simply remain with the stdlib's classical `OrderedType` as far as functor arguments are concerned.

That is `Pqueue.mli` would look like this:

```ocaml

module type OrderedType = sig type t val compare : t -> t -> int end

(* Priority on elements *)

module type Min = sig type t type elt … end
module type Max = sig type t type elt … end
module MakeMin (Elt : OrderedType) : Min with type elt = Elt.t
module MakeMax (Elt : OrderedType) : Max with type elt = Elt.t

(* Separate priorities, polymorphic elements *)

module type MinPriority = sig type 'a t type priority … end
module type MaxPriority = sig type 'a t type priority … end
module MakeMinPriority (Priority : OrderedType) : MinPriority with type priority = Priority.t
module MakeMaxPriority (Priority : OrderedType) : MaxPriority with type priority = Priority.t
```

This also shows that we could also simply start only with `Make{Min,Max}` and act on `Make{Min,Max}Priority` only the day a person shows up with a compelling use case on the issue tracker :-)

backtracking commented:
Many thanks @gasche and @dbuenzli for your feedback on this PR!
I'll be happy to make any change once we converge on some API.
Notes:
* I'm perfectly fine leaving out the `Int` instances.
* I like the last proposal with `MakeMinPriority` and `MakeMaxPriority`, which is simpler than the current `MakePoly`.

dbuenzli commented:
As far as I'm concerned, I think the `Min/Max` interfaces should be rather uncontroversial, I had a quick look and it seems the names you choose are `Stdlib` consistent.

There is likely a bit to bikeshed on the `{Min,Max}Priority` ones. I see a few options:

1. You define  `type 'a {Min,Max}Priority.elt = priority * 'a` and you basically make it exactly like `Min/Max`. This is likely the simplest and most consistent. 
2. You allow some currying e.g.: 
   ```
   add : 'a t -> priority -> 'a -> unit`
   iter : (priority -> 'a -> unit) -> 'a t -> unit
   …
   ```
   But then in some places you have the tuples anyway (e.g. `of_list`, etc.).
3. If you don't do 1. should you return the priorities when you `max_elt` or `pop_max` or only the `'a` value ?

Note that since you will store the elements according to 1. there's no (non-)allocation argument to be had, the tuple will be created. So perhaps the simplest, consistent and requiring less discussion :–) is to choose 1.

dbuenzli commented:
Another advantage of 1. in my last message is that, case arising, it makes it easier to switch code from one variant to the other.

gasche commented:
One problem that I see in practice with the API that @dbuenzli proposed in https://github.com/ocaml/ocaml/pull/12871#issuecomment-2094258140 is that it involves a lot of duplication of interfaces:
- `Min` and `Max` are almost the same; this can possibly be avoided by defining a signature for their intersection and including it, or maybe defining several shared subsignatures to have more control over the ordering in the generated documentation.
- `Min` and `MinPriority` are also almost the same, and mostly canot be shared.
 
Duplicating interfaces is bad, because they contain documentation comments, and we don't want to duplicate those.

(There is also a little terminology issue, the module is called "PQueue" like "Priority Queues" but in fact it exposes Sorted/Ordered Queues, only some of which have a separate notion of priority. I propose to keep P/Priority in the module name as this is the well-known name of the data structure.)

I think it would be better to offer just one interface for min queues, or offer several but with a mechanical way to derive them from a single one. Ideally we could write:

```ocaml
module type OrderedQeue = sig
  type 'a t
  val add : 'a t -> 'a -> unit
  ...
end

module type PriorityQueue = sig
  module Elem : OrderedType
  type t
  include OrderedQueue with type 'a t = t constraint 'a = Elem.t
end
```

but the module language does not support this.

dbuenzli commented:
> Duplicating interfaces is bad, because they contain documentation comments, and we don't want to duplicate those.

I have to categorically reject this argument :–) This is ignoring the awesome advances made by `odoc` team. 

This [gist](https://gist.github.com/dbuenzli/22c177c7e4daaae389b1013f5344acd3) shows how to structure things so that 1) **no** documentation comment whatsoever gets duplicated 2) The [rendered documentation](https://erratique.ch/tmp/pqueue/html/pqueue/Pqueue/index.html) shows none of the sharing cruft (we could even put the hidden interfaces in a submodule so that they don't show too much on tab completions).

I think that https://github.com/ocaml/ocaml/pull/12871#issuecomment-2094258140 is a nice API for users to use (personally the subtle distinction between priority queues and sorted queues doesn't bother me more than that).

gasche commented:
Thanks! I had the odoc `@inline` annotation in mind to share interface mixins between Min and Max, but I didn't think of obtaining the monomorphic and polymorphic signatures by instantiating the same generalized interface. This is quite the expertise.

With this in mind I don't have an objection to the proposed signature. I would have a slight preference for exposing `Make{Min,Max}Poly` instead of `Make{Min,Max}Priority`, because I think that it reduces the risk of users accidentally using `MakeMinPriority` when `MakeMin`. (In any case I agree with your suggestion of literally having `type 'a elt = priority * 'a` in the Priority signature instead of trying to curry things around; I like the consistency.)

backtracking commented:
Thanks again for your comments and suggestions.

I'm also fine with `type 'a elt = priority * 'a` and the idea of giving up with curried functions. After all, pairs are used internally in any case.

@gasche I don't like much the terminology "ordered queue", as this is not what we find in the literature.

backtracking committed:
stdlib: yet another interface for priority queues

we now have polymorphic (min and max) priority queues, with separate
priorities, and monomorphic ones, with priority on elements

backtracking commented:
Hi @gasche and @dbuenzli ,

Many thanks once again for your detailed reviews and suggestions.

I have adopted (mostly) the proposal of @dbuenzli . Here are the main changes:
* polymorphic priority queues with separate priorities `MakeMinPriority` and `MakeMaxPriority`
* monomorphic priority queues with priority on elements `MakeMin` and `MakeMax`
* no more currying
* no more integer instances (as it is now straightforward to do `MakeMin(Int)`)

Notes:
* I deliberately chose to keep duplicated signatures for `Min` and `Max`. I've seen the suggestion by @dbuenzli involving lots of `include`, and the related comment about `odoc`, but I find it much simpler and cleaner to have it like this. Note that it is done like this in other modules such as `Hashtbl` (where the same signature appears three times).
* The documentation of `Dynarray` still contains an implementation of priority queues as a code example. I can remove it. Just ask.

dbuenzli commented:
Thanks @backtracking. Re-reading five days later our discussion with @gasche on the [gist here](https://gist.github.com/dbuenzli/22c177c7e4daaae389b1013f5344acd3) I think I'm starting to bias towards [his proposal](https://gist.github.com/dbuenzli/22c177c7e4daaae389b1013f5344acd3?permalink_comment_id=5053390#gistcomment-5053390). 

Initially I reasoned thinking the issue was about separate priorities or not. But now that we no longer try to curry it becomes obvious that it's not the point, those can equally be created with `Make{Min,Max}`, which uses the regular `OrderedType` argument. The point is about how easy we want it for users to be able to create polymorphic queues. 

As I said at the end of the discussion, my impression is that creating polymorphic queues is going to be the exception rather than the rule and that in this case his design, while a little bit more mind twisting, is slightly more general and less ad-hoc (namely the use of pairs in `{Max,Min}Priority`).

Perhaps other people should chime in on what they think about that aspect. I think however that the base `Make{Min,Max}`, everyone should be fine with.

Sorry if this comment comes out as frustrating, but additions to the `Stdlib` tend to be forever so we need to get that right :–)

fpottier commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,217 @@
+        minimal priority, or returns [None] if the queue is empty. *)
+
+    val remove_min: 'a t -> unit
+    (** [pop_min_opt q] removes an element in queue [q] with
```
Typo: the comment mentions `pop_min_opt` instead of `remove_min`.

backtracking replied:
Thanks! Fixed.

backtracking commented:
> Thanks @backtracking. Re-reading five days later our discussion with @gasche on the gist I think I'm starting to bias towards his proposal [...] Sorry if this comment comes out as frustrating, but additions to the `Stdlib` tend to be forever so we need to get that right :–)

No problem. I don't mind going for a fourth API. Since the implementation is unique and never changes, that's no big deal. I'm on it...

backtracking committed:
stdlib: a fourth API for priority queues

we still have both monomorphic and polymorphic functors,
but the polymorphic functors do not have separate priorities anymore

code-ghalib commented:
I wonder if we can get rid of the `Min/Max` distinction, let the comparator decide the order, and call the "best" element the `top` of the queue? So only the functors `Make` and `MakePoly` will be left? (btw, do we also intend to add `MakePoly` to `Set` or `Map`?)

backtracking commented:
> I wonder if we can get rid of the `Min/Max` distinction, let the comparator decide the order, and call the "best" element the `top` of the queue? 

Thanks for the feedback @code-ghalib !
Initially, I did that, but it happened confusing (what is the meaning of "best"? is this "min" or "'max"? and so on). See the discussion above. So at some point I switched to the two interfaces `Min` and `Max`, which I find much nicer now. As pointed out in https://github.com/ocaml/ocaml/pull/12871/#issuecomment-1891508200, this is also what is done in other libraries.

Please note that there is a single implementation. The four functors are only different ways of presenting it to the user. No code is duplicated.

code-ghalib commented:
> Thanks for the feedback @code-ghalib ! Initially, I did that, but it happened confusing (what is the meaning of "best"? is this "min" or "'max"? and so on). See the discussion above. So at some point I switched to the two interfaces `Min` and `Max`, which I find much nicer now. As pointed out in #12871, this is also what is done in other libraries.

That's a fair concern. It also looks similar to the [haskell interface](https://hackage.haskell.org/package/pqueue). I was just comparing to other languages, where the terms [`top`](https://en.cppreference.com/w/cpp/container/priority_queue/top), [`pop`](https://github.com/python/cpython/blob/3.12/Lib/heapq.py#L137), [`peek`](https://docs.rs/priority-queue/latest/priority_queue/priority_queue/struct.PriorityQueue.html#method.peek) are used to get around having to pick a direction for "best".  _The top of the queue_ is also an expression I've heard english speakers use often. There are also others that have just picked either min or [max](https://pkgs.racket-lang.org/package/priority-queue) with the implicit assumption that users will know how to reverse the priority.

> 
> Please note that there is a single implementation. The four functors are only different ways of presenting it to the user. No code is duplicated.  

Yes, I just wish that there was less multiplicity in the presentation - and the complexity feels a bit incidental tbh, as opposed to essential to the problem. I considered proposing that we move the comparator dispatch to runtime, and then even the distinction for polymorphic elements goes away, so there's no functors at all:
```ocaml
type 'a t
val create : compare:('a -> 'a -> int) -> unit -> 'a t
```  

But I figured this might face more opposition - it's not very much like `Map` or `Set` where we use functors.

gasche commented:
I think that the sense of confusion we have about what "best" means is something, as API designers, is something that our user would also feel. It would make them have to double-check the documentation and maybe test to know whether the "min" or "max" element is at the top of the queue. (Especially since there is a difference between the typical computer-science convention, where the *smallest* element is chosen first, and the typical normal-life convention where having "priority 10" means that you are chosen before "priority 2". To avoid this confusion, some of the other APIs you cited deviate from it and return the "largest" element by default (C++, Rust), but other stick to the smallest element (your Python example)... making things even more confusing!) It is not pleasant at all to be unsure how something is going to behave and to be forced to read the manual (gasp!).

Having both Min and Max means a little bit more complexity in the stdlib implementation, but I think that it actually makes it easier for our user to understand how to use the interface, and it makes them easier to read the code using the structure. I like the fact that we shifting the costs from our users to ourselves.

backtracking commented:
>  I considered proposing that we move the comparator dispatch to runtime, and then even the distinction for polymorphic elements goes away, so there's no functors at all:
> 
> ```ocaml
> type 'a t
> val create : compare:('a -> 'a -> int) -> unit -> 'a t
> ```
> 
> But I figured this might face more opposition - it's not very much like `Map` or `Set` where we use functors.

Interestingly, it used to be like that a long time ago in Caml Light, when there were no functors yet, if I remember correctly. But then you have to store the `compare` function inside the data structure and you can't serialize anymore. Functors are a better approach for this reason (I think).

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    [min_elt_opt]) is the one that is removed from the priority queue
+    by [pop_min] (or [pop_min_opt] or [remove_min]). This is important
+    in many algorithms, (e.g. when peeking at several priority queues
+    and then selecting one to remove from).
```
My question when reading this is rather: is there any order guarantees (LIFO, FIFO, none) on pop if I insert many elements with the same priority.

backtracking replied:
There is no guarantee (such as LIFO or FIFO) when you have many elements with the same priority. There is not enough structure internally to provide it (this is related to other discussions below, such as the order of iteration for instance).

I improved the documentation to make it more explicit (see 6d558c74).

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+
+    exception Empty
+    (** Raised when {!min_elt}, {!pop_min} or {!remove_min} is applied
+        to an empty queue. *)
```
I do not want to open a long debate on that but I still think the question should be asked whether we want to have a design with exceptions here. Regardless of any consistency argument with the `Queue` module: there's no use in being consistent when one is wrong :-)

We know the perils of using exceptions for non exceptional events, we know an empty queue is not exceptional, we know it's harder to reason about exceptions and we know that providing choice slows down programmers (also AFAIR in many scenarios, whether using exceptions is really faster is quite inconclusive).

If I was the person designing this API I would just provide this:

```ocaml
val min_elt : t -> elt option
val pop_min : t -> elt option
val remove_min : t -> unit (* Doesn't matter if [t] is empty. *)
```
Short, to the point, explicit, no stack footguns, no conditions to check. 

Possibly adding: 

```ocaml
val get_min_elt : t -> elt (* Raises [Invalid_argument] if [t] is empty. *)
```
but since nowadays we have a much better stdlib that's just `Q.min_elt q |> Option.get` which I find as nice to read and since I'm sure one day the gods of optimisation will manage to optimize that away to the equivalent implementation, likely not worth adding to the API.

backtracking replied:
I would be fine with `min_elt` returning an option, and `get_min_elt` raising an exception, but I dislike the lack of uniformity with the remaining of the stdlib we would get. I have mixed feelings...

dbuenzli replied:
Note that `get*` functions raising `Invalid_argument` if they can't handle the request is rather consistently done in the stdlib, e.g. `{String,Bytes,Array,Dynarray}.get*`, `Option.get`, `Result.get_*`. But as mentioned I don't think it's really necessary to add it. My hunch here is rather on the definition of a new exception for an unexceptional condition.

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    val min_elt_opt: t -> elt option
+    (** [min_elt_opt q] returns an element in queue [q] with minimal priority,
+        without removing it from the queue, or returns [None] if the queue
+        is empty. *)
```
I find this doc string noisy. I would prefer a declarative doc string here. Namely: 
```ocaml
val min_elt_opt: t -> elt option
(** [min_elt_opt q] is an element of [q] with minimal priority or [None] 
    if [q] is empty. *)
```

backtracking replied:
Fixed

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+        minimal priority, or raises {!Empty} if the queue is empty. *)
+
+    val clear: t -> unit
+    (** Discard all elements from a priority queue. *)
```
Since the function is `remove_min` I would rather like to keep that terminology in the doc string: 

```
(** [clear q] removes all elements from [q]. *)
```

dbuenzli replied:
And if the doc string is changed to what I suggested, the doc strings talks backs to the identifier and the question naturally comes on whether this should not simply be called `remove_all`. One less verb to learn.

backtracking replied:
Fixed.

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+
+    val copy: t -> t
+    (** [copy q] is a shallow copy of [q] (a new priority queue
+        containing the same elements as [q]). *)
```
Since `deep/shallow` copy is not really widespread in OCaml terminology I would drop it and simply keep the parenthesised definition which not much longer: 

```
(** [copy q] is a new priority queue with the same elements [q] has. *)
```

backtracking replied:
Fixed.

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    (** [add q x] adds the element [x] in the priority queue [q]. *)
+
+    val add_seq: t -> elt Seq.t -> unit
+    (** [add q s] adds the elements of [s] in the priority queue [q]. *)
```
The usual debate: why not `add_list`, `add_array`, `add_dynarray`, `add_set`, `add_map`,  `add_zorglub`. Going through `Seq.t` for this is wasteful. The usual functional programming answer: 

```ocaml
val add_iter : t -> ((elt -> unit) -> 'v -> unit) -> 'v -> unit
(** [add_iter q iter v] adds each element of [v] to [q]. This is 
    [iter (add q) v]. *)
```

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    (** {1:iteration Iteration}
+
+        The order in which the elements of a priority queue are
+        traversed is unspecified.
```
My uses of priority queues never needed such functions. But I think it's worth asking: is that expected ? It looks suprising to me. My gut feeling when reading `Pqueue.fold` would be that I'm being folded in priority order. 

I think if you want to keep these *names*, these traversals should be made ordered, that's the reason d'être of priority queues. If you don't want to implement ordered traversals and keep these functions, I would suggest to prefix them with `unordered_`. Simply dropping these functions is yet another option, but I can see them as being useful.

backtracking replied:
Hi,
Thanks for the feedback.
It is not possible to traverse keys in increasing order. There is not enough structure for that internally: indeed, this is a heap data structure (encoded within an array) and not a binary search tree. So we know where the smallest element is (the root) but the other elements are scattered in the subtrees in a way which is not at all sorted.
Said otherwise, traversing elements in increasing order would be as costly as extracting them one by one (with `pop`), with a total cost O(N log N).

backtracking replied:
Now, calling the function `iter_unsorted` (and `fold_unsorted`) to stress this fact could be an option indeed. I'm a bit reluctant because all other iteration functions in the stdlib are just called `iter`.

In the meantime, I improved the documentation string of `iter` and `fold`.

dbuenzli replied:
> I'm a bit reluctant because all other iteration functions in the stdlib are just called `iter`.

That's precisely the reason why I'm very reluctant with the current name :-) All these `iter` functions guarantee a traversal order that is natural for the underlying structure. So if I read `Pqueue.iter` I'm also going to assume that. If I read `Pqueue.unordered_iter`, I'm not going to. 

I think I rather have the `unordered` as prefix here, it reads more naturally. Regarding `unsorted` vs `unordered` I don't know but the surrounding context is all about `order` so `unordered` feels more natural to me.

Octachron replied:
Note that there is already `Hashtbl.iter` which has an unspecified order of iteration.

dbuenzli replied:
Right but there's no real notion of order between members of a hash table. Besides it does have a least recently defined notion of order for those bindings that are specified more than once.

Meanwhile `Queue`, `Stack`, `Map`, `Set`, `List`, `Array`, `Dynarray`, `String`, `Bytes` all have a notion of order and their `iter` function respects it. 

Now `Pqueue` has a notion of order and its `iter` function does not respect it, for me that's asking for misunderstandings (a.k.a. introducing a footgun).

gasche replied:
For the record, I'm with @dbuenzli on that one, this is something that we discussed with @backtracking and I proposed `iter_unordered`, which has my preference over just `iter`. I would also be okay with `unordered_iter`. I think that calling just the function `iter` does not respect the principle of least surprise, and having user reach for a differently-named function will make the surprising behavior much more clear to them.

gasche replied:
(One nice thing with `iter_unordered` is that it will show up as a completion proposition if people write `iter`.)

backtracking replied:
Thanks for the discussion. I'll go for `iter_unordered` and `fold_unordered` then.

backtracking replied:
Done

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    val of_seq: elt Seq.t -> t
+    (** [of_seq s] returns a new priority queue containing the
+        elements of sequence [s]. Runs in linear time. *)
+
```
Mirroring `add_iter`, `of_iter` should be added:

```ocaml
val of_iter : ((elt -> unit) -> 'v -> unit) -> 'v -> unit
(** [of_iter iter v] is a new priority queue containing the elements of [v]. *)
```

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    The following, more complex functors create polymorphic queues of
+    type ['a t], just like other polymorphic containers (lists,
+    arrays...). They require a notion of "polymorphic elements" ['a
+    elt] that can be compared without depending on choice of ['a].
```
depending on choice of ['a] -> without depending on the values of ['a].

backtracking replied:
Fixed

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+        type 'a elt = task
+        let compare t1 t2 = Int.compare t1.priority t2.priority
+      end)
+    ]}
```
Could we perhaps try to find a good example for that ? Or failing to simply drop this example: this `'a task` is [totally unconvincing](https://gist.github.com/dbuenzli/22c177c7e4daaae389b1013f5344acd3?permalink_comment_id=5053464#gistcomment-5053464) to me. I think the separate priorities that follows is sufficient to justify the poly signatures.

backtracking replied:
Thanks for the feedback. I removed this first example scenario. (in commit 6d558c74)

dbuenzli commented on stdlib/pqueue.mli:
```diff
@@ -0,0 +1,299 @@
+    val to_seq: t -> elt Seq.t
+    (** [to_seq q] is the sequence of elements of [q] in some
+        unspecified order. *)
+
```
Not necessarily asking for but it. But should we add the usual `equal` and (priority aware) `compare` functions ?

backtracking replied:
Hi,
Thanks for this comment.
It is not possible to provide `equal` or `compare`, as there is no canonical structure inside (and generally speaking, not enough information to retrieve keys in order; see above). Said otherwise, it would be way too costly to attempt at providing `equal` or `compare` i.e. as costly as extracting all elements one by one.

dbuenzli reviewed (changes requested):
If people are still around this summer I think it would be nice to try to get this in for 5.3 (which gets freezed mid-august). So many more OCaml 5 schedulers have yet to be written and would benefit from this!

I made a review round on the API on which I still have a few nags. Once we settle these I'm willing to do a full review of the implementation.

dbuenzli commented:
So for 5.4 perhaps ? @backtracking

backtracking commented:
My bad. I thought I was done with this PR and I missed the fact it was not making its way to 5.3.

gasche commented:
Apologies, I should have been more explicit, but there hasn't been a full review of the current iteration of the PR yet, and @dbuenzli's review comments should be adressed. (It's okay to disagree with some of them and not make changes for those, hopefully with an explanation.)

(I am not worried about the implementation review, I believe that I looked at it previously and was convinced by its correctness. The hard part is to agree on the API.)

backtracking committed:
various small changes in Pqueue interface

add_seq/of_seq replaced by add_iter/of_iter
improved documentation string at various places

backtracking commented:
Hi @dbuenzli ,
Thanks again for the review and the feedback. I tried to address all your points in the last commit 6d558c74.

backtracking committed:
adding a suffix _unordered to iter and fold

and removing to_seq in addition

gasche commented:
@dbuenzli when you have time, I wonder if you have a global opinion on the current state of this PR. Are you broadly satisfied, what are the remaining sticking points? My intention is to approve once the discussion has stabilized to a compromise that both you and @backtracking are happy with.

My impression is that the only remaining sticking point is the matter of exceptions or options (and naming) on functions that extract the first element.

dbuenzli commented:
> My impression is that the only remaining sticking point is the matter of exceptions or options (and naming) on functions that extract the first element.

Yes. I would prefer to see it reduced to [that](https://github.com/ocaml/ocaml/pull/12871#discussion_r1687798935), but then I'm also a bit tired of being the only one who seems to be always complaining about this. So if I'm the only one what can I say ? I don't want to block the addition on that but I don't understand why people find it useful to inflict on users of the stdlib this complexity in terms of learning, choice, control flow behaviour and reasoning (not to mention the internal code duplication). 

I also don't see the rationale behind `Q.remove_min` throwing an exception if the queue is empty. All of `{Hashtbl,Set,Map}.remove`, `List.remove_{assq,assoc}`, `Dynarray.remove_last`, `Weak.remove`, `Ephemeron.remove` are just a no-op if the requested element to remove is not present. I'm not sure why it should be different for queues. For me this is again a violation of least surprise w.r.t. an established name of the stdlib.

(Also if `Q.remove_min` is not made to raise. One can decently ask if it's useful to introduce an `Empty` exception just for `min_elt` and `pop_min`. One is just making the API harder to learn. With an option, it's self-evident, you won't even need a trip to the docs).

gadmm commented:
@dbuenzli's suggestion of
```ocaml
val min_elt : t -> elt option
val pop_min : t -> elt option
val get_min_elt : t -> elt (* Raises [Invalid_argument] if [t] is empty. *)
```
seems to be the recommendation for best practice regarding exceptions in interfaces

gasche commented:
@dbuenzli I wasn't trying to suggest that you are blocking this PR, on the contrary I think that the result is nicer now that what we started from. I find it helpful to have you as a sort of extremal point in opinionated-ness: if Daniel approves an API, then surely I can approve as well, and if Daniel only half-approves I can check the remaining open points to see if they feel acceptable.

On this specific point of exceptions I don't have a strong opinion myself, I can hear arguments for your proposal and also one for doing this "the old way". My approach is to push the decision burden to @backtracking: if he finds himself clearly convinced by either approaches, I would be inclined to go with it, giving deference to the author of the whole module.

backtracking commented:
Thanks @dbuenzli and @gasche for your comments!

Let me clarify a first point: the *only* reason why I introduced an `Empty` exception is consistency with other modules of the standard library, notably `Stack` and `Queue`. But I must admit that I kind of dislike such `Empty` exceptions myself, so, since you are happy not to be consistent with the past, I'll get rid of it.

Second, for `option` return types instead of an exception, this is slightly different: I could invoke consistency, but these days we have `option` variants as well, and it seems that OCaml users are fine with using options instead of exceptions for such things. For long, I thought it was really a pity to allocate where there is no need to allocate, but I guess the OCaml compiler made enough progress these days for the options not to be that costly. So fine, I'll go for `min_elt` returning an option (together with `get_min_elt` raising `Invalid_argument`).

Last, for `remove_min`, this is another matter. For me, this is a fundamental operation of the data structure, similar to `pop` on a stack. I would find *really weird* that it stays silent if the priority queue is empty. (I don't see it similar to a nonexistent binding in a dictionary.) So I won't make the change (but I'll move `Empty` to `Invalid_argument`).

Stay tuned for the commit implementing these changes...

backtracking committed:
priority queues: no more Empty exception

in addition, min_elt now returns an option (and get_min_elt raises
an exception, namely Invalid_argument), as well as pop_min

dbuenzli commented:
> Last, for `remove_min`, this is another matter. For me, this is a fundamental operation of the data structure, similar to `pop` on a stack. I would find _really weird_ that it stays silent if the priority queue is empty.

Then I'd suggest you use another name like `delete_min`. But note that:

1. `Dynarray.remove_last` is morally equivalent to `pop` on a stack but does not raise.
2. You can also say that `Q.pop_min` is fundamental and `Q.remove_min q` is just `Q.pop_min q |> ignore`.
3. Since `Invalid_argument` denotes programming error and is not supposed to be triggered by programmers this means that either a) you know by invariant that the queue is not empty and you can invoke `Q.delete_min` directly or b) that you always need to invoke `Q.is_empty` (or `Q.min_elt`, or `Q.length`) before invoking `Q.delete_min`.

Regarding 3. while you may be sometimes making anyways the `Q.is_empty` check as part of a loop, I personally find the extra check more painful *in general*. 

When it makes sense, being able to invoke operations on data structures regardless of their state, without having to check conditions beforehand is stricly more usable to me. Like for `Dynarray.remove_last` there's nothing wrong in doing nothing if there's nothing to remove here, it's just the equivalent of a logical vacuous truth.

gasche commented:
Note: if there is a disagreement on `remove_{min,max}` (which is not an essential function), a safe way out is just to remove it.

backtracking commented:
> Note: if there is a disagreement on `remove_{min,max}` (which is not an essential function), a safe way out is just to remove it.

I guess you mean in favor of `|> ignore` then. So I have to choose between two solutions I dislike (`remove_min` doing nothing and `pop_min` unnecessarily building an option to be ignored). You guys are tough negotiators!

I go for `remove_min` doing nothing.

backtracking committed:
priority queues: remove_min does nothing if the queue is empty

dbuenzli reviewed (approved):
>  Functions [sift_up] and [sift_down] do not perform swaps, but
rather expect the value to be assigned in the heap as an additional parameter [x], resulting in twice less assignments.

Nice, my implementations never used that trick.

I have reviewed the `.mli` and `.ml` and as far as I'm concerned this is good to go.

redianthus commented on winpthreads:
Is this change expected? (there's also a similar `flexdll` that has been deleted)

gasche reviewed (approved):
I'm approving the PR.

( @zapashcanon noticed that two submodules had been deleted in the PR, this is a mistake that must be reverted. To be fair, submodules are fairly obscure to work with.)

gasche commented:
This needs the approval of another maintainer to be merged.

Octachron commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,287 @@
+
+    (* We say that a heap respects the "heap ordering" if the value of
+       each node is no greater than the value of its children. The
+       algorithm manipulates arrays that respect the heap algorithm,
```
heap algorithm ⇒ heap ordering

backtracking replied:
Thanks for spotting the typo. it's fixed.

Octachron commented on stdlib/pqueue.ml:
```diff
@@ -0,0 +1,287 @@
+      h
+
+    let of_array a =
+      Dynarray.of_array a |> heapify
```
Small remark: as far as I can see, the current implement could support an "efficient" inplace map as
```ocaml
let map f x = heapify (Dynarray.map_inplace f x)
```
(if `Dynarray.map_inplace` existsted) which cannot be implemented from the interface. I don't know however if that function would be worth exposing?

backtracking replied:
Thanks for the feedback. Indeed, we could have such a `map_inplace` (if we have `Dynarray.map_inplace`) but note that it would have the same asymptotic complexity as building a fresh array with the mapped values and then use `of_array`. It would only save us the intermediate array, and be more efficient by a small constant factor.

Octachron reviewed (approved):
The new module seems useful to me, and after review its interface feels sufficiently minimalist to leave room for potential future evolutions (in both the interface and the implementation).

I am approving, but the testsuite needs to be fixed and the submodules restored.

backtracking commented:
> I am approving, but the testsuite needs to be fixed and the submodules restored.

Thanks @Octachron !

Regarding the testsuite, I've added a new file `testsuite/tests/lib-pqueue/test.ml`, which I'm running with success as follows:
```
$ ./ocamltest/ocamltest testsuite/tests/lib-pqueue/test.ml 
 ... testing 'test.ml' => passed
```
Is there anything else I should be doing regarding the test suite?

lthls commented:
> Is there anything else I should be doing regarding the test suite?

The CI is failing for Windows workers because of the deleted submodules. I assume that once you've fixed that the testsuite will run fine.

backtracking commented:
> The CI is failing for Windows workers because of the deleted submodules. I assume that once you've fixed that the testsuite will run fine.

Thanks. I have no idea what these submodules are and how to fix them. Any help is welcome.

nojb commented:
> > The CI is failing for Windows workers because of the deleted submodules. I assume that once you've fixed that the testsuite will run fine.
> 
> Thanks. I have no idea what these submodules are and how to fix them. Any help is welcome.

@backtracking: executing
```
git checkout 744e158697c570005b8047c6a315d63259617ea8 -- flexdll winpthreads
```
and committing the result should fix the issue.

backtracking commented:
Thanks @nojb !

gasche commented:
Are you okay with me squashing the commits together, or do you want to cleanup the history manually?

gasche commented:
@backtracking I want to click "merge" but there are two things, a minor and a less minor one:

- The "Hygiene" CI check is failing because the versioned .depend file is now out of date; you should run `make depend` (or maybe `make alldepend`) and commit the result to .depend files. This is minor and I was willing to merge and fix it myself after the merge.
- Last month @zapashcanon asked whether `remove_min` should be called `drop_min`, because `Stack` has `drop`. There was a tense battle around the semantics of `remove_min` but we forgot to bike-shed over the name. Last chance to delay a nice PR with inevitable discussions, @backtracking and @dbuenzli what do you think about `remove_min` versus `drop_min`? (`delete_min` was also briefly mentioned)

Let me quote @zapashcanon:

> Maybe it could be named `drop_min` ? So that we would have `Pqueue.{pop_min,drop_min}` like in `Stack.{pop,drop}`. `Queue` only has `pop`, I opened #12884 to add `drop`.

dbuenzli commented:
> * @backtracking and @dbuenzli what do you think about `remove_min` versus `drop_min`? (`delete_min` was also briefly mentioned)

`drop*` functions seem to raise on empty structures. So given the current semantics of `remove_min`, it's [the right](https://github.com/ocaml/ocaml/pull/12871#issuecomment-2634236795) name.

backtracking commented:
Hi @gasche ,
I updated the `.depend`. For `remove_min`, let us keep the name, if you don't mind, since @dbuenzli and myself are both happy with it.

gasche commented:
I had started doing the `alldepend` myself on my machine. I ended up cleaning your commit history manually -- I prefer this to squashing as it reflects the non-trivial amount of work that went in this PR. I also rebased on top of the current trunk, and just force-pushed the result to your branch -- so please reset to `backtracking/stdlib-priority-queues` first before making new changes. I hope that there will be no new changes to do! This is good to merge if the CI is happy.

backtracking commented:
Thanks @gasche !

dbuenzli commented:
Thanks @backtracking! 

It turns out I have an immediate need for priority queues, this will prove handy :-)

nojb commented:
@backtracking sorry to come late to the discussion, but recently I tried to use this module and I realized it is missing the "remove" operation. Since how to implement this operation (and the related `change_priority` operation) is nicely explained in the exercises of `Learn Programming with OCaml` https://usr.lmf.cnrs.fr/lpo/, I was wondering why you chose to omit them from the API in this PR. Thanks!

backtracking commented:
Hi @nojb ,
Thanks for this comment. Supporting `remove` and `change_priority` requires a more complex structure and a more complex API, with handles to elements. It uses more space, loses the good data locality, exposes the user to the risk of manipulating handles not valid anymore, etc.

My feeling is that if you need them, then you'll implement your own priority queues. Possibly using a book  😉

nojb commented:
OK, thanks!
